applications/nrf5340_audio/src/modules/lc3_streamer.c

```diff
@@ -0,0 +1,412 @@
+
+	ret = data_fifo_pointer_first_vacant_get(&stream->fifo, (void **)&data_ptr, K_NO_WAIT);
+	if (ret) {
+		LOG_ERR("Failed to get first vacant block %d", ret);
+		return ret;
+	}
+
+	ret = lc3_file_frame_get(&stream->file, data_ptr,
+				 CONFIG_SD_CARD_LC3_STREAMER_MAX_FRAME_SIZE);
+	if (ret) {
+		LOG_ERR("Failed to get frame from file %d", ret);
+		data_fifo_block_free(&stream->fifo, (void *)data_ptr);
+		return ret;
+	}
+
+	ret = data_fifo_block_lock(&stream->fifo, (void **)&data_ptr,
+				   CONFIG_SD_CARD_LC3_STREAMER_MAX_FRAME_SIZE);
+	if (ret) {
+		LOG_ERR("Failed to lock block %d", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Loop the stream by closing and re-opening the file, and loading the first frame.
+ *
+ * @param[in]	stream	Pointer to the stream to loop.
+ *
+ * @retval	0	Success, negative value otherwise.
+ */
+static int loop_stream(struct lc3_stream *stream)
+{
+	int ret;
+
+	ret = lc3_file_close(&stream->file);
+	if (ret) {
+		LOG_ERR("Failed to close file %d", ret);
+		return ret;
+	}
+
+	ret = lc3_file_open(&stream->file, stream->filename);
+	if (ret) {
+		LOG_ERR("Failed to open file %d", ret);
+		return ret;
+	}
+
+	ret = put_next_frame_to_fifo(stream);
+	if (ret) {
+		LOG_ERR("Failed to put first frame after loop to fifo %d", ret);
+
+		int lc3_file_ret = lc3_file_close(&stream->file);
+
+		if (lc3_file_ret) {
+			LOG_ERR("Failed to close file %d", lc3_file_ret);
+		}
+
+		return ret;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Load the next frame from the stream to the fifo. This is the work queue function.
+ *
+ * @param[in]	work	Pointer to the work queue item.
+ */
+static void load_next_frame(struct k_work *work)
+{
+	int ret;
+	struct lc3_stream *stream = CONTAINER_OF(work, struct lc3_stream, work);
+
+	ret = put_next_frame_to_fifo(stream);
+	if (ret == -ENODATA) {
+		LOG_DBG("End of stream");
+		if (stream->loop_stream) {
+			ret = loop_stream(stream);
+			if (ret) {
+				LOG_ERR("Failed to loop stream %d", ret);
+				stream->state = STREAM_ENDED;
+			}
+		} else {
+			stream->state = STREAM_PLAYING_LAST_FRAME;
+		}
+	} else if (ret) {
+		LOG_ERR("Failed to put next frame to fifo %d", ret);
+		stream->state = STREAM_ENDED;
+	}
+}
+
+int lc3_streamer_next_frame_get(int streamer_idx, uint8_t **frame_buffer)
+{
+	int ret;
+	char *data_ptr;
+	size_t data_len;
+
+	if (streamer_idx < 0 || streamer_idx >= CONFIG_SD_CARD_LC3_STREAMER_MAX_NUM_STREAMS) {
+		LOG_ERR("Invalid streamer index %d", streamer_idx);
+		return -EINVAL;
+	}
+
+	struct lc3_stream *stream = &streams[streamer_idx];
+
+	if ((stream->state != STREAM_PLAYING) && (stream->state != STREAM_PLAYING_LAST_FRAME)) {
+		LOG_ERR("Stream not playing");
+		return -EFAULT;
+	}
+
+	if (stream->active_buffer != NULL) {
+		data_fifo_block_free(&stream->fifo, (void *)stream->active_buffer);
+		stream->active_buffer = NULL;
+	}
+
+	if (stream->state == STREAM_PLAYING_LAST_FRAME) {
+		LOG_INF("Stream ended");
+		stream->state = STREAM_ENDED;
+		return -ENODATA;
```

Comment:
The idea is that `-ENODATA` will signal the caller that there is no more data to be read, and that the stream should be closed with `lc3_streamer_end_stream()`. We can't call `lc3_streamer_end_stream()` here directly, because we need to keep the data in `frame_buffer` valid until the caller is done reading it.